lib/fluent/plugin/in_tail/group_watch.rb

```diff
@@ -0,0 +1,204 @@
+        group_watcher
+      end
+
+      def remove_path_from_group_watcher(path)
+        return if @group.nil?
+        group_watcher = find_group_from_metadata(path)
+        group_watcher.delete(path)
+      end
+
+      def construct_group_key(named_captures)
+        match_rule = []
+        @group_keys.each { |key|
+          match_rule.append(named_captures.fetch(key, GroupWatchParams::DEFAULT_KEY))
+        }
+        match_rule = match_rule.join(GroupWatchParams::REGEXP_JOIN)
+
+        match_rule
+      end
+
+      def construct_groupwatchers
+        @group.rule.each { |rule|
+          match_rule = construct_group_key(rule.match)
+          @group_watchers[match_rule] ||= GroupWatcher.new(@group.rate_period, rule.limit)
+        }
+      end
+
+      def find_group(metadata)
+        metadata_key = construct_group_key(metadata)
+        gw_key = @group_watchers.keys.find { |regexp| metadata_key.match?(regexp) && regexp != @default_group_key }
+        gw_key ||= @default_group_key
+
+        @group_watchers[gw_key]
+      end
+
+      def find_group_from_metadata(path)
+        begin
+          metadata = @group.pattern.match(path).named_captures
+          group_watcher = find_group(metadata)
+        rescue
+          log.warn "Cannot find group from metadata, Adding file in the default group"
+          group_watcher = @group_watchers[@default_group_key]
+        end
+
+        group_watcher
+      end
+    end
+
+    class GroupWatcher
+      attr_accessor :current_paths, :limit, :number_lines_read, :start_reading_time, :rate_period
+
+      FileCounter = Struct.new(
+        :number_lines_read,
+        :start_reading_time,
+      )
+
+      def initialize(rate_period = 60, limit = -1)
+        @current_paths = {}
+        @rate_period = rate_period
+        @limit = limit
+      end
+
+      def add(path)
+        @current_paths[path] = FileCounter.new(0, nil)
+      end
+
+      def include?(path)
+        @current_paths.key?(path)
+      end
+
+      def size
+        @current_paths.size
+      end
+
+      def delete(path)
+        @current_paths.delete(path)
+      end
+
+      def update_reading_time(path)
+        @current_paths[path].start_reading_time ||= Fluent::Clock.now
+      end
+
+      def update_lines_read(path, value)
+        @current_paths[path].number_lines_read += value
+      end
+
+      def reset_counter(path)
+        @current_paths[path].start_reading_time = nil
+        @current_paths[path].number_lines_read = 0
+      end
+
+      def time_spent_reading(path)
+        Fluent::Clock.now - @current_paths[path].start_reading_time
+      end
+
+      def limit_time_period_reached?(path)
+        time_spent_reading(path) < @rate_period
+      end
+
+      def limit_lines_reached?(path)
+        return true unless include?(path)
+        return true if @limit == 0
+
+        return false if @limit < 0
+        return false if @current_paths[path].number_lines_read < @limit / size
```

Comment:
Thanks for adding this great feature!
I have a question about this code line.

The `@limit` controls the total number of lines collected for a group, right? (https://docs.fluentd.org/input/tail#less-than-group-greater-than-section)

That judgment seems to be done by this code line, but this line judges it by the average number of lines collected for a group, not the total number.

Either the code or the documentation is wrong?

Reply:
Hmm, that's right. It should be total.

Reply:
Thanks! For now, I made an issue for this: 

* #4184

Reply:
@daipom @ashie 
>This PR enables us to form different groups based on configuration rules provided by the user. Within each group, one can set the rate limits for log collection. The set rate limit gets uniformly applied to each member (log file) of the group. 

The PR description states that the limit set in the configuration distributes the rate uniformly to all the members in a group. That's why the condition checks for average rate limit for a particular group member. I get why this is confusing and we may either change the documentation with an example or change the code itself. 

The reason we are using average limit (for a group member) is because it puts a hard upper limit on the rate of logs emitted by a group. For example, a two groups with same limit but one having 1000 containers and the other having 10 containers, both will emit the same number of log lines. 
However, if you apply the same limit to all the group members (without having a hard upper limit control) then there's no limit on the amount of data that is sent via fluentd. Considering the previous example, the rate of data sent from one group will be 100x larger.

Reply:
@Pranjal-Gupta2 Thanks!

> Considering the previous example, the rate of data sent from one group will be 100x larger.

If we apply the limit to each file independently, it will be 100 times larger, right?

I think the preferred specification is to calculate the total number in the group in #4184.
In this way, it will be the same upper amount independent of the number of files.

Reply:
>If we apply the limit to each file independently, it will be 100 times larger, right? 

Yes, correct! 

> I think the preferred specification is to calculate the total number in the group in https://github.com/fluent/fluentd/issues/4184.
> In this way, it will be the same upper amount independent of the number of files.

Can you please elaborate on this?

Reply:
> I think the preferred specification is to calculate the total number in the group in https://github.com/fluent/fluentd/issues/4184.
> In this way, it will be the same upper amount independent of the number of files.

Are you saying that we calculate the limit per file once and then use it for rate limiting for each group member, right? Taking average ensures that if a file is being added/removed dynamically, then rate limit adjusts itself.

Reply:
I'm thinking of having just one counter per group and comparing the total as a group with the limit.

Code image (This is just an image of a modification):

```diff
--- a/lib/fluent/plugin/in_tail/group_watch.rb
+++ b/lib/fluent/plugin/in_tail/group_watch.rb
@@ -138,17 +138,18 @@ module Fluent::Plugin
       )
 
       def initialize(rate_period = 60, limit = -1)
-        @current_paths = {}
+        @current_paths = []
         @rate_period = rate_period
         @limit = limit
+        @counter = FileCounter.new(0, nil)
       end
 
       def add(path)
-        @current_paths[path] = FileCounter.new(0, nil)
+        @current_paths << path
       end
 
       def include?(path)
-        @current_paths.key?(path)
+        @current_paths.include?(path)
       end
 
       def size
@@ -159,25 +160,25 @@ module Fluent::Plugin
         @current_paths.delete(path)
       end
 
-      def update_reading_time(path)
-        @current_paths[path].start_reading_time ||= Fluent::Clock.now
+      def update_reading_time
+        @counter.start_reading_time ||= Fluent::Clock.now
       end
 
-      def update_lines_read(path, value)
-        @current_paths[path].number_lines_read += value
+      def update_lines_read(value)
+        @counter.number_lines_read += value
       end
 
-      def reset_counter(path)
-        @current_paths[path].start_reading_time = nil
-        @current_paths[path].number_lines_read = 0
+      def reset_counter
+        @counter.start_reading_time = nil
+        @counter.number_lines_read = 0
       end
 
-      def time_spent_reading(path)
-        Fluent::Clock.now - @current_paths[path].start_reading_time
+      def time_spent_reading
+        Fluent::Clock.now - @counter.start_reading_time
       end
 
-      def limit_time_period_reached?(path)
-        time_spent_reading(path) < @rate_period
+      def limit_time_period_reached?
+        time_spent_reading < @rate_period
       end
 
       def limit_lines_reached?(path)
@@ -185,13 +186,13 @@ module Fluent::Plugin
         return true if @limit == 0
 
         return false if @limit < 0
-        return false if @current_paths[path].number_lines_read < @limit / size
+        return false if @counter.number_lines_read < @limit
 
         # update_reading_time(path)
-        if limit_time_period_reached?(path) # Exceeds limit
+        if limit_time_period_reached? # Exceeds limit
           true
         else # Does not exceed limit
-          reset_counter(path)
+          reset_counter
           false
         end
       end
```

Reply:
Does this implementation ensure that the fluentd **collects equally** from all the files in a group?

Reply:
Oh! I see!
Now I can see the essence of this.
We can prevent unbalanced collections by judging the limit with the average number.
As you say, simply judging by the total number may result in unbalanced collections.

We need to think more carefully about this feature's specifications.
Thanks!

Reply:
Yup! We can write a better articulation of this point in the documentation. 
@ashie Can I take this up?

Reply:
I will think some more about what to do with #4184.
If it looks good to fix the document, I will fix it.

Reply:
> Yup! We can write a better articulation of this point in the documentation. @ashie Can I take this up?

Oh, sorry. I missed this comment.
At this point, fixing the document would be a good idea!

Reply:
We welcome the document modifications of course!

Reply:
Thanks @daipom @ashie! 
Will start working on this.